bc_obps/Dockerfile

```diff
@@ -1,58 +1,109 @@
-# Use an official Python runtime as a parent image
-FROM python:3.12.3 AS main
+####################
+# Base Stage
+####################
+FROM python:3.12.3-slim AS base
 
-# Install system dependencies and clean up
-RUN apt-get update && \
-    apt-get upgrade -y && \
-    apt-get install -y git gnupg curl build-essential && \
-    apt-get clean
-
-# Set environment variables
+# Prevent Python from writing .pyc files to disk
 ENV PYTHONDONTWRITEBYTECODE=1 \
+# Ensure Python output is sent straight to terminal (no buffering)
     PYTHONUNBUFFERED=1 \
+# Define a user ID for the non-root user
     USER_ID=1001 \
-    HOME=/root
+# Set the home and working directory to /app
+    HOME=/app \
+# Prevent apt-get from prompting during installs
+    DEBIAN_FRONTEND=noninteractive
 
 WORKDIR ${HOME}
-# Make everything in the home group-writable to support OpenShift's restricted SCC
-# Needs to be done as root to chown
-RUN useradd -ms /bin/bash -d ${HOME} --uid ${USER_ID} -g root bc_obps
 
-COPY ./ ${HOME}/
+# Recreate bc_obps user with bash shell
+RUN useradd -ms /bin/bash -d ${HOME} --uid ${USER_ID} -g root bc_obps && \
+    chown -R bc_obps:0 ${HOME} && \
+    chmod -R g+rwX ${HOME}
+
+####################
+# Builder Stage
+####################
+FROM base AS builder
+
+# Refresh package lists from repositories
+RUN apt-get update && \
+# Install minimal build tools required for asdf (git), curl for Poetry plugin, build-essential for compiling Python, and additional libs for Python
+    apt-get install -y --no-install-recommends build-essential curl git libc6-dev libffi-dev libssl-dev zlib1g-dev && \
+# Remove apt cache to reduce image size
+    apt-get clean && \
+# Delete package lists to further slim the image
+    rm -rf /var/lib/apt/lists/*
 
-# Install asdf
-RUN git clone https://github.com/asdf-vm/asdf.git ${HOME}/asdf --depth 1 --branch v0.11.2
+# Copy version config and Poetry files with ownership set to bc_obps
+COPY --chown=bc_obps:0 .tool-versions pyproject.toml poetry.lock ${HOME}/
 
+USER ${USER_ID}
+
+# Clone asdf shallowly from GitHub
+RUN git clone --depth 1 --branch v0.11.2 https://github.com/asdf-vm/asdf.git ${HOME}/asdf
```

Comment:
```suggestion
RUN git clone --depth 1 --branch v0.16.2 https://github.com/asdf-vm/asdf.git ${HOME}/asdf
```

Reply:
This has some breaking changes, Is that fine?

```
NOTICE:

You have tried to upgrade to asdf 0.16.0 or newer. Versions 0.16.0 is a
complete rewrite of asdf in Go. This text is being printed by the older
Bash implementation. If you are seeing this you have not migrated to
asdf 0.16.0. Please follow the instructions on the upgrade guide to
migrate to the new version.

Aside from this notice, this older Bash implementation works as it did
in asdf version 0.15.0 and older.

Migration guide: https://asdf-vm.com/guide/upgrading-to-v0-16
asdf website: https://asdf-vm.com
Source code: https://github.com/asdf-vm/asdf
```

Reply:
https://asdf-vm.com/guide/upgrading-to-v0-16